File: php/tests/NumericConstraintsTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\ErrorCode;
use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * §6.1, §9.3 — Numeric constraints (integer and number kinds).
 */
class NumericConstraintsTest extends TestCase
{
    use TestHelpers;

    private function errs(mixed $value, string $kind = 'number', array $constraints = []): array
    {
        $r = new Registry();
        $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/num-test',
            'types' => [
                'Num' => ['properties' => ['v' => array_merge(['type' => $kind], $constraints)]],
            ],
        ]);
        $schema = $r->getSchema('https://example.org/schemas/num-test');
        return validate(['_type' => 'Num', 'v' => $value], $schema->types['Num'], $schema, $r);
    }

    public function test_minimum_ok(): void
    {
        $this->assertSame([], $this->errs(5, constraints: ['minimum' => 0]));
    }

    public function test_minimum_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs(-1, constraints: ['minimum' => 0]), ErrorCode::BELOW_MINIMUM));
    }

    public function test_maximum_ok(): void
    {
        $this->assertSame([], $this->errs(10, constraints: ['maximum' => 10]));
    }

    public function test_maximum_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs(11, constraints: ['maximum' => 10]), ErrorCode::ABOVE_MAXIMUM));
    }

    public function test_exclusive_minimum_ok(): void
    {
        $this->assertSame([], $this->errs(1, constraints: ['exclusiveMinimum' => 0]));
    }

    public function test_exclusive_minimum_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs(0, constraints: ['exclusiveMinimum' => 0]), ErrorCode::BELOW_EXCLUSIVE_MINIMUM));
    }

    public function test_exclusive_maximum_ok(): void
    {
        $this->assertSame([], $this->errs(9, constraints: ['exclusiveMaximum' => 10]));
    }

    public function test_exclusive_maximum_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs(10, constraints: ['exclusiveMaximum' => 10]), ErrorCode::ABOVE_EXCLUSIVE_MAXIMUM));
    }

    public function test_multiple_of_ok(): void
    {
        $this->assertSame([], $this->errs(6, constraints: ['multipleOf' => 3]));
    }

    public function test_multiple_of_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs(7, constraints: ['multipleOf' => 3]), ErrorCode::NOT_MULTIPLE_OF));
    }

    public function test_integer_ok(): void
    {
        $this->assertSame([], $this->errs(3, kind: 'integer'));
    }

    public function test_integer_float_with_fraction(): void
    {
        $this->assertTrue($this->hasCode($this->errs(3.5, kind: 'integer'), ErrorCode::NOT_INTEGER));
    }

    public function test_integer_float_no_fraction(): void
    {
        // 3.0 is acceptable as integer (no fractional part)
        $this->assertSame([], $this->errs(3.0, kind: 'integer'));
    }

    public function test_enum_number_ok(): void
    {
        $this->assertSame([], $this->errs(2, constraints: ['enum' => [1, 2, 3]]));
    }

    public function test_enum_number_fail(): void
    {
        $this->assertTrue($this->hasCode($this->errs(5, constraints: ['enum' => [1, 2, 3]]), ErrorCode::ENUM_MISMATCH));
    }
}

File: java/src/test/java/org/oojs/NumericConstraintsTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class NumericConstraintsTest extends TestHelpers {

    private List<ValidationError> errs(Object value, String kind, Map<Object, Object> constraints) {
        Registry r = new Registry();
        Map<Object, Object> propDef = map("type", kind);
        propDef.putAll(constraints);
        r.loadMap(map("$oojs", "1.0", "$id", "https://example.org/schemas/num-test",
                "types", map("Num", map("properties", map("v", propDef)))));
        Schema schema = r.getSchema("https://example.org/schemas/num-test");
        return new Validator(r).validate(map("_type", "Num", "v", value), schema.types.get("Num"), schema);
    }

    private List<ValidationError> errs(Object value, Map<Object, Object> constraints) {
        return errs(value, "number", constraints);
    }

    @Test void minimum_ok()              { assertEquals(List.of(), errs(5, map("minimum", 0))); }
    @Test void minimum_fail()            { assertTrue(hasCode(errs(-1, map("minimum", 0)), ErrorCode.BELOW_MINIMUM)); }
    @Test void maximum_ok()              { assertEquals(List.of(), errs(10, map("maximum", 10))); }
    @Test void maximum_fail()            { assertTrue(hasCode(errs(11, map("maximum", 10)), ErrorCode.ABOVE_MAXIMUM)); }
    @Test void exclusive_minimum_ok()    { assertEquals(List.of(), errs(1, map("exclusiveMinimum", 0))); }
    @Test void exclusive_minimum_fail()  { assertTrue(hasCode(errs(0, map("exclusiveMinimum", 0)), ErrorCode.BELOW_EXCLUSIVE_MINIMUM)); }
    @Test void exclusive_maximum_ok()    { assertEquals(List.of(), errs(9, map("exclusiveMaximum", 10))); }
    @Test void exclusive_maximum_fail()  { assertTrue(hasCode(errs(10, map("exclusiveMaximum", 10)), ErrorCode.ABOVE_EXCLUSIVE_MAXIMUM)); }
    @Test void multiple_of_ok()          { assertEquals(List.of(), errs(6, map("multipleOf", 3))); }
    @Test void multiple_of_fail()        { assertTrue(hasCode(errs(7, map("multipleOf", 3)), ErrorCode.NOT_MULTIPLE_OF)); }
    @Test void integer_ok()              { assertEquals(List.of(), errs(3, "integer", map())); }
    @Test void integer_float_fraction()  { assertTrue(hasCode(errs(3.5, "integer", map()), ErrorCode.NOT_INTEGER)); }
    @Test void integer_float_no_fraction(){ assertEquals(List.of(), errs(3.0, "integer", map())); }
    @Test void enum_number_ok()          { assertEquals(List.of(), errs(2, map("enum", list(1, 2, 3)))); }
    @Test void enum_number_fail()        { assertTrue(hasCode(errs(5, map("enum", list(1, 2, 3))), ErrorCode.ENUM_MISMATCH)); }
}
